### PR TITLE
Enable strict mode in tsconfig.json

### PR DESCRIPTION
**Pull request by j0nas**
# Enable strict mode in tsconfig.json

## Intent

Enable TypeScript _strict mode_ for _/web_.

Solves the underlying issue of [this Shortcut Story](https://app.shortcut.com/sanity-io/story/19097/urgent-site-not-refreshing).

## Description

The `strict` setting in _tsconfig.json_ was set to `false`, which opts us out of a lot of the benefits that one comes to expect of a compiled, type-safe language. Unfortunately, I didn't discover this until quite recently, when the linked Shortcut Story was reported. This PR enables this setting for the _web_ directory, and does the bare minimum necessary to have CI passing again with the new checks in place. [As noted on Slack](https://wja.slack.com/archives/C02AX66E2S0/p1651479109756419), having this flag set to `true` would likely have prevented the build error encountered in the linked Story.

This PR touches many files, but you will quickly discover that most of the fixes are fairly similar in nature. My approach to enable the `strict` flag was to fix the reported issues one-by-one until there were none left. This may have led to some types of fixes being applied in a slightly inconsistent way. This was a result of the sheer number of errors initially reported, and the fact that I didn't discover that I could report _all_ errors using `npx tsc --noEmit` (instead of one-by-one as was the output of `yarn build`) until quite late into the fixing process. (I had to stop myself several times from performing larger-scale refactorings; such things should definitely be done in follow-up issues & PRs.) Sorry if this PR feels a bit chaotic. Please feel free to suggest fixes for any such inconsistencies if you find any, or (preferably) file issues that are to be followed up in later PRs.

## Testing this PR

Verify that the [Preview Deploy for this PR](https://structured-content-2022-web-git-tsconfig-strict-mode.sanity.build/) renders identically to other Preview Deploys using the `staging` dataset (e.g. the [Preview Deploy](https://structured-content-2022-web-git-generate-og-image.sanity.build/) for #167). Expecting to go through every single page is not realistic, but focus on the most important parts, for instance the front page and the Program page.

### CODE REVIEW

**Review by oyvind-stenhaug — CHANGES_REQUESTED** [review]
Great work, I'm glad TS can be made to catch a few more issues than it previously has.

Added some issues and comments inline, though most of them are labelled as nit. Might be worth noting that GitHub hides some of it behind a "Load more" button.

File: web/components/SessionSpeakers/SessionSpeakers.tsx [hunk 1]
```diff
@@ -60,7 +62,7 @@ export const SessionSpeakers = ({
       {speaker2 && <Speaker speaker={speaker2} variant="nonDesktop" />}
       <Shape />
     </div>
-    <div className={styles.column2} aria-hidden={speaker2 ? null : 'true'}>
+    <div className={styles.column2} aria-hidden={Boolean(speaker2)}>
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/SessionSpeakers/SessionSpeakers.tsx]: dd7761b.

File: web/components/Nav/MenuItem/MenuItem.tsx [hunk 2]
```diff
@@ -14,14 +14,17 @@ export const MenuItem = ({
...
+  const Anchor = (
+    <a
+      {...rest}
+      className={clsx(styles.link, href === currentPath && styles.current)}
+      onClick={closeMenu}
+    />
+  );
+  return (
+    <li className={className}>
+      {href ? <Link href={href}>{Anchor}</Link> : Anchor}
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/Nav/MenuItem/MenuItem.tsx]: 82ceee9.

File: web/components/TextBlock/SponsorsSection/SponsorsSection.tsx [hunk 3]
```diff
@@ -59,12 +59,15 @@ export const SponsorsSection = ({
...
+            .filter(
+              (sponsorship) =>
+                sponsorship.sponsors?.length && sponsorship.sponsors.length > 0
+            )
             .map(({ _id, type, sponsors }) => (
               <section key={_id} className={styles.sponsorLevel}>
                 <Heading type="h3">{type}</Heading>
                 <ul className={styles.sponsors}>
-                  {sponsors.map((sponsor) => (
+                  {sponsors?.map((sponsor) => (
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/TextBlock/SponsorsSection/SponsorsSection.tsx:70]: Yeah. Predicates are a bit clumsy to define when dealing with complex/nested types, so I figured that the optional chaining was a simpler way to go about it to appease TS.

File: web/components/TextBlock/Programs/Programs.tsx [hunk 4]
```diff
@@ -27,71 +33,70 @@ type ProgramsProps = {
...
-                          .saturation(-100)
-                          .url()}
-                        width={40}
-                        height={40}
-                        alt={person.photo.alt || ''}
-                      />
-                    )}
+                {speakers.map(({ person }) => {
+                  const personPhotoSrc =
+                    person?.photo &&
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/TextBlock/Programs/Programs.tsx]: 7bb71f2

File: web/components/TextBlock/Programs/Programs.tsx [hunk 5]
```diff
@@ -152,40 +166,38 @@ export const Programs = ({
...
-                    <SessionSection
-                      {...{ session, activeProgram, start }}
-                      key={session._key}
-                    />
-                  );
-                })}
+              {sessionsPerDay[day].filter(hasSession).map((session) => {
+                const start = sessionStart(
+                  activeProgram.startDateTime,
+                  session.session!._id,
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/TextBlock/Programs/Programs.tsx]: 95295a7

File: web/components/TextBlock/Programs/Programs.tsx [hunk 6]
```diff
@@ -152,40 +166,38 @@ export const Programs = ({
...
+                );
+                return shouldLinkToSession(session.session) ? (
+                  <Link
+                    href={getEntityPath(session.session)}
+                    key={session._key}
+                  >
+                    <a className={styles.sessionLink}>{Section}</a>
+                  </Link>
+                ) : (
+                  Section
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/TextBlock/Programs/Programs.tsx]: 51acbc4.

File: web/components/TextBlock/Sponsorships/Sponsorships.tsx [hunk 7]
```diff
@@ -66,7 +67,11 @@ export const Sponsorships = ({
             </thead>
             <tbody>
               {allBenefitNamesSortOrderPreserved(allSponsorships).map((b) => (
-                <BenefitRow key={b} name={b} sponsorships={sponsorships} />
+                <BenefitRow
+                  key={b}
+                  name={b || ''}
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/TextBlock/Sponsorships/Sponsorships.tsx]: 4b8a2b9? Feels slightly more idiomatic.

File: web/components/VenueNav/VenueNav.tsx [hunk 8]
```diff
@@ -14,7 +14,7 @@ interface VenueNavProps {
 export const VenueNav = ({ venues, activeVenue, mainVenue }: VenueNavProps) => {
   const venueLabel = (venueName: string) =>
     ({
-      [mainVenue?.name]: 'Main venue',
+      [mainVenue?.name || '']: 'Main venue',
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/components/VenueNav/VenueNav.tsx]: d879c53

File: web/pages/[[...slug]].tsx [hunk 9]
```diff
@@ -249,11 +249,12 @@ export const getStaticPaths: GetStaticPaths = async () => {
...
 export const getStaticProps: GetStaticProps = async ({
-  params: { slug: slugParam },
+  params,
   preview = false,
 }) => {
+  const slugParam = params?.slug?.[0] ?? '';
   const slug = Array.isArray(slugParam)
-    ? urlJoin.apply(null, [...slugParam, { leadingSlash: false }])
+    ? slugParam.reduce((acc, cv) => urlJoin(acc, cv, { leadingSlash: false }))
     : slugParam || '/';
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/pages/[[...slug]].tsx]: Yeah, looking at this with fresh eyes, I can immediately tell this doesn't look right. In 3df84d3, I copied over a `getSlug` util function which I initially wrote for the [Venue Details](https://github.com/sanity-io/structured-content-2022/compare/venue-details) branch. which saves us duplicating these possibly confusing snippets. What do you think?

**Comment by j0nas** [comment]
@oyvind-stenhaug Thank you for the thorough review. :100:

File: web/components/SessionSpeakers/SessionSpeakers.tsx [hunk 10]
```diff
@@ -60,7 +62,7 @@ export const SessionSpeakers = ({
       {speaker2 && <Speaker speaker={speaker2} variant="nonDesktop" />}
       <Shape />
     </div>
-    <div className={styles.column2} aria-hidden={speaker2 ? null : 'true'}>
+    <div className={styles.column2} aria-hidden={!Boolean(speaker2)}>
```
Comment (oyvind-stenhaug) [inline web/components/SessionSpeakers/SessionSpeakers.tsx:65]: Hm, I was about to write that this is probably fine. But now I see this note: https://w3c.github.io/aria/#h-note-53

> At the time of this writing, `aria-hidden="false"` is known to work inconsistently in browsers. As future implementations improve, use caution and test thoroughly before relying on this approach.

So I would really prefer to keep the original behavior of not outputting the attribute at all when the speaker is absent. As far as I can tell, TS is appeased by replacing the original `null` with `undefined`:

```suggestion
    <div className={styles.column2} aria-hidden={speaker2 ? undefined : true}>
```

File: web/components/TextBlock/Sponsorships/Sponsorships.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (oyvind-stenhaug) [reply web/components/TextBlock/Sponsorships/Sponsorships.tsx]: Looks OK to me, the renaming of `b` is also good

File: web/components/VenueNav/VenueNav.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (oyvind-stenhaug) [reply web/components/VenueNav/VenueNav.tsx]: OK. Switch is not capable of skipping the main venue comparison if there is no main venue, but as mentioned that's a corner case anyway.

File: web/util/pages.ts [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+import urlJoin from 'proper-url-join';
+
+export const getSlug = (params?: ParsedUrlQuery) => {
+  const slug = params?.slug;
+  if (!slug) {
+    return '';
+  }
+
+  return Array.isArray(slug)
+    ? slug.reduce((acc, curr) => urlJoin(acc, curr, { trailingSlash: false }))
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/util/pages.ts]: Mea culpa, good catch.

File: web/util/pages.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { ParsedUrlQuery } from 'querystring';
```
[the remark this answers is not in the document]
Comment (j0nas) [reply web/util/pages.ts]: bc2de6b

**Review by oyvind-stenhaug — APPROVED** [review]
Thanks, looks good to me!

**Pull request merged**